aplib.net-demo/Assets/Scripts/WFC/Grid.cs

```diff
@@ -76,5 +111,119 @@ public void PlaceRoom(int x, int y, Room room)
+        /// <seealso cref="GetConnectedNeighbours"/>
+        public ICollection<Cell> Get8NeighbouringCells(Cell cell)
+        {
+            ICollection<Cell> neighbours = new Collection<Cell>();
+            for (int i = -1; i < 2; i++)
+            {
+                if (cell.X + i < 0 || cell.X + i >= Width) continue; // No out of range
+                for (int j = -1; j < 2; j++)
+                {
+                    if (i == j && i == 0) continue; // Skip the cell itself
+                    if (cell.Y + j < 0 || cell.Y + j >= Height) continue; // No out of range
+                    neighbours.Add(this[cell.X + i, cell.Y + j]);
+                }
+            }
+            return neighbours;
+        }
+        
+        /// <summary>
+        /// Given a cell, this method determines all directly adjacent neighbours which are connected by their <see cref="Tile"/>s.
+        /// These are at most 4, as diagonal tiles cannot be connected.
+        /// </summary>
+        /// <param name="cell">The cell whose connected tiles are to be determined.</param>
+        /// <returns>The cells directly connected through their tiles</returns>
+        /// <remarks>Assumes that the cells are assigned a tile.</remarks>
+        public ICollection<Cell> GetConnectedNeighbours(Cell cell)
+        {
+            ICollection<Cell> connectedNeighbours = new Collection<Cell>();
+            ICollection<Cell> neighbours = Get4NeighbouringCells(cell); // Note: no diagonal neighbours
+            foreach (Cell neighbour in neighbours)
+            {
+                if      (cell.Tile.CanConnectInDirection(1) && neighbour.X > cell.X && neighbour.Tile.CanConnectInDirection(3))
+                    connectedNeighbours.Add(neighbour);
+                else if (cell.Tile.CanConnectInDirection(3) && neighbour.X < cell.X && neighbour.Tile.CanConnectInDirection(1))
+                    connectedNeighbours.Add(neighbour);
+                else if (cell.Tile.CanConnectInDirection(0) && neighbour.Y > cell.Y && neighbour.Tile.CanConnectInDirection(2))
+                    connectedNeighbours.Add(neighbour);
+                else if (cell.Tile.CanConnectInDirection(2) && neighbour.Y < cell.Y && neighbour.Tile.CanConnectInDirection(0))
+                    connectedNeighbours.Add(neighbour);
+            }
+
+            return connectedNeighbours;
+        }
+
+        /// <summary>
+        /// This method goes through all cells in this grid, to determine all connected components.
+        /// </summary>
+        /// <returns>A collection of sets of cells, where the sets of cells represent the connected components.</returns>
+        public IList<ISet<Cell>> DetermineConnectedComponents()
+        {
+            ISet<Cell> unvisitedCells = new HashSet<Cell>(_cells.Where(cell => cell.Tile is not Empty)); // Deep copy
+            IList<ISet<Cell>> connectedComponents = new List<ISet<Cell>>();
+
+            while (unvisitedCells.Any())
```

Comment:
Set implements ICollection and thus has a property Count. Use this instead of the slow Linq call.

Reply:
And here starts an interesting debate. Count is a property, but that does not exclude that on first 'get', everything needs to be counted.

In contrast, any() iterates the collection and stops when the first item has been found.

I will dive into the Count property to discover what its complexity is

Reply:
Count is updated when calling functions so I think it is also set when initializing the list. Therefore the first 'get' is O(1).

Reply:
Indeed. As far as I know, count is updated on `Add` and `Remove` (and their families of methods). When constructing a collection, it is done using a series of Adds (or something like it), and will thus update the count property.

As for Any(). This calls upon the Enumerator, which is known to be slow, even if it should only iterate once. Hence, I would argue `Count` is the better property.

I don't understand why a property would need to be calculated? Which datastructure does this, and why would it be implemented this way? Did you perhaps confuse it with the LINQ method `Count()`?

Reply:
Count is probably low-complexity. COunt it is!